File: apps/mobile/ios-native/BridgeProfileStore.swift
import CryptoKit
import Foundation

final class BridgeProfileStore {
    private let defaults = UserDefaults.standard

    private let preferencesName = "codex.nativeHost.profiles"
    private let activeProfileKey = "codex.nativeHost.activeProfileID"
    private let legacyNameKey = "codex.nativeHost.legacy.name"
    private let legacyEndpointKey = "codex.nativeHost.legacy.serverEndpoint"
    private let legacyAuthTokenKey = "codex.nativeHost.legacy.authToken"

    func list() -> [BridgeProfile] {
        let rawProfiles = defaults.string(forKey: preferencesName)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !rawProfiles.isEmpty {
            return persistSanitizedProfilesIfNeeded(parseProfiles(rawProfiles))
        }
        return migrateLegacyProfile().map { [$0] } ?? []
    }

    func readActive() -> BridgeProfile? {
        let profiles = list()
        guard !profiles.isEmpty else {
            return nil
        }
        let activeProfileID = defaults.string(forKey: activeProfileKey)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return profiles.first(where: { $0.id == activeProfileID }) ?? profiles.first
    }

    func write(_ profile: BridgeProfile) {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let profileToSave = BridgeProfile(
            id: profile.id,
            bridgeID: profile.bridgeID,
            name: profile.name,
            serverEndpoint: profile.serverEndpoint,
            authToken: profile.authToken,
            lastUsedAtMilliseconds: profile.lastUsedAtMilliseconds ?? now
        )

        var profiles = list()
        profiles.removeAll { $0.id == profileToSave.id }
        profiles.insert(profileToSave, at: 0)
        persistProfiles(profiles, activeProfileID: profileToSave.id)
    }

    func setActive(_ profileID: String) {
        var profiles = list()
        guard let activeProfile = profiles.first(where: { $0.id == profileID }) else {
            return
        }
        let refreshed = BridgeProfile(
            id: activeProfile.id,
            bridgeID: activeProfile.bridgeID,
            name: activeProfile.name,
            serverEndpoint: activeProfile.serverEndpoint,
            authToken: activeProfile.authToken,
            lastUsedAtMilliseconds: Int64(Date().timeIntervalSince1970 * 1000)
        )
        profiles.removeAll { $0.id == profileID }
        profiles.insert(refreshed, at: 0)
        persistProfiles(profiles, activeProfileID: refreshed.id)
    }

    @discardableResult
    func remove(_ profileID: String) -> BridgeProfile? {
        let currentProfiles = list()
        let remainingProfiles = currentProfiles.filter { $0.id != profileID }
        let currentActiveID = defaults.string(forKey: activeProfileKey)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let nextActiveID: String?
        if remainingProfiles.isEmpty {
            nextActiveID = nil
        } else if currentActiveID == profileID {
            nextActiveID = remainingProfiles.first?.id
        } else if remainingProfiles.contains(where: { $0.id == currentActiveID }) {
            nextActiveID = currentActiveID
        } else {
            nextActiveID = remainingProfiles.first?.id
        }
        persistProfiles(remainingProfiles, activeProfileID: nextActiveID)
        return remainingProfiles.first(where: { $0.id == nextActiveID })
    }

    func clear() {
        defaults.removeObject(forKey: preferencesName)
        defaults.removeObject(forKey: activeProfileKey)
    }

    func createProfileID(name: String, endpoint: String, bridgeID: String? = nil) -> String {
        let seed = "\(bridgeID?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "")|\(name.trimmingCharacters(in: .whitespacesAndNewlines))|\(endpoint.trimmingCharacters(in: .whitespacesAndNewlines))|\(Date().timeIntervalSince1970)"
        let digest = SHA256.hash(data: Data(seed.utf8))
        let suffix = digest.prefix(6).map { String(format: "%02x", $0) }.joined()
        return "bridge_\(Int(Date().timeIntervalSince1970))_\(suffix)"
    }

    private struct ParsedProfiles {
        let profiles: [BridgeProfile]
        let mutated: Bool
    }

    private func parseProfiles(_ rawProfiles: String) -> ParsedProfiles {
        guard let data = rawProfiles.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data, options: []) as? [JSONDictionary] else {
            return ParsedProfiles(profiles: [], mutated: true)
        }

        var mutated = false
        let profiles = root.compactMap { item -> BridgeProfile? in
            let id = (item["id"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let endpoint = (item["serverEndpoint"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !id.isEmpty, !endpoint.isEmpty else {
                mutated = true
                return nil
            }
            if let legacyPayload = (item["tailnetEnrollmentPayload"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
               !legacyPayload.isEmpty {
                mutated = true
            }

            return BridgeProfile(
                id: id,
                bridgeID: (item["bridgeId"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
                name: (item["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? endpoint,
                serverEndpoint: endpoint,
                authToken: (item["authToken"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
                lastUsedAtMilliseconds: (item["lastUsedAtMillis"] as? Int64) ?? (item["lastUsedAtMillis"] as? Int).map(Int64.init)
            )
        }
        return ParsedProfiles(profiles: profiles, mutated: mutated)
    }

    private func persistSanitizedProfilesIfNeeded(_ parsed: ParsedProfiles) -> [BridgeProfile] {
        guard parsed.mutated else {
            return parsed.profiles
        }
        let currentActiveID = defaults.string(forKey: activeProfileKey)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let nextActiveID = parsed.profiles.first(where: { $0.id == currentActiveID })?.id ?? parsed.profiles.first?.id
        persistProfiles(parsed.profiles, activeProfileID: nextActiveID)
        return parsed.profiles
    }

    private func persistProfiles(_ profiles: [BridgeProfile], activeProfileID: String?) {
        let serializedProfiles = profiles.map { profile in
            [
                "id": profile.id,
                "bridgeId": profile.bridgeID as Any,
                "name": profile.name,
                "serverEndpoint": profile.serverEndpoint,
                "authToken": profile.authToken as Any,
                "lastUsedAtMillis": profile.lastUsedAtMilliseconds as Any,
            ] as JSONDictionary
        }
        defaults.set(jsonEncodedString(serializedProfiles), forKey: preferencesName)
        defaults.set(activeProfileID, forKey: activeProfileKey)
        defaults.removeObject(forKey: legacyNameKey)
        defaults.removeObject(forKey: legacyEndpointKey)
        defaults.removeObject(forKey: legacyAuthTokenKey)
    }

    private func migrateLegacyProfile() -> BridgeProfile? {
        let endpoint = defaults.string(forKey: legacyEndpointKey)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !endpoint.isEmpty else {
            return nil
        }
        let name = defaults.string(forKey: legacyNameKey)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? endpoint
        let authToken = defaults.string(forKey: legacyAuthTokenKey)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        let profile = BridgeProfile(
            id: createProfileID(name: name, endpoint: endpoint),
            bridgeID: nil,
            name: name,
            serverEndpoint: endpoint,
            authToken: authToken,
            lastUsedAtMilliseconds: Int64(Date().timeIntervalSince1970 * 1000)
        )
        persistProfiles([profile], activeProfileID: profile.id)
        return profile
    }
}

File: apps/mobile/android/app/src/main/java/com/boomyao/codexmobile/nativehost/BridgeProfileStore.kt
package com.boomyao.codexmobile.nativehost

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.abs

class BridgeProfileStore(context: Context) {
    private val preferences =
        context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    fun list(): List<BridgeProfile> {
        val rawProfiles = preferences.getString(KEY_PROFILES_JSON, null)?.trim().orEmpty()
        if (rawProfiles.isNotEmpty()) {
            return persistSanitizedProfilesIfNeeded(parseProfiles(rawProfiles))
        }
        return migrateLegacyProfile()?.let(::listOf).orEmpty()
    }

    fun readActive(): BridgeProfile? {
        val profiles = list()
        if (profiles.isEmpty()) {
            return null
        }
        val activeProfileId = preferences.getString(KEY_ACTIVE_PROFILE_ID, null)?.trim().orEmpty()
        return profiles.firstOrNull { it.id == activeProfileId } ?: profiles.firstOrNull()
    }

    fun write(profile: BridgeProfile) {
        val now = System.currentTimeMillis()
        val profileToSave = profile.copy(lastUsedAtMillis = profile.lastUsedAtMillis ?: now)
        val profiles = list().toMutableList()
        profiles.removeAll { it.id == profileToSave.id }
        profiles.add(0, profileToSave)
        persistProfiles(profiles, profileToSave.id)
    }

    fun setActive(profileId: String) {
        val profiles = list().toMutableList()
        val activeProfile = profiles.firstOrNull { it.id == profileId }?.copy(lastUsedAtMillis = System.currentTimeMillis()) ?: return
        profiles.removeAll { it.id == activeProfile.id }
        profiles.add(0, activeProfile)
        persistProfiles(profiles, activeProfile.id)
    }

    fun remove(profileId: String): BridgeProfile? {
        val currentProfiles = list()
        val removedProfile = currentProfiles.firstOrNull { it.id == profileId }
        val remainingProfiles = currentProfiles.filterNot { it.id == profileId }
        val currentActiveId = preferences.getString(KEY_ACTIVE_PROFILE_ID, null)?.trim().orEmpty()
        val nextActiveId =
            when {
                remainingProfiles.isEmpty() -> null
                currentActiveId == profileId -> remainingProfiles.first().id
                remainingProfiles.any { it.id == currentActiveId } -> currentActiveId
                else -> remainingProfiles.first().id
            }
        persistProfiles(remainingProfiles, nextActiveId)
        return remainingProfiles.firstOrNull { it.id == nextActiveId }
    }

    fun clear() {
        preferences.edit().clear().apply()
    }

    fun createProfileId(name: String, endpoint: String, bridgeId: String? = null): String {
        val seed = "${bridgeId?.trim().orEmpty()}|${name.trim()}|${endpoint.trim()}|${System.currentTimeMillis()}"
        return "bridge_${System.currentTimeMillis()}_${abs(seed.hashCode())}"
    }

    private data class ParsedProfiles(
        val profiles: List<BridgeProfile>,
        val mutated: Boolean,
    )

    private fun parseProfiles(rawProfiles: String): ParsedProfiles {
        return runCatching {
            val root = JSONArray(rawProfiles)
            var mutated = false
            buildList {
                for (index in 0 until root.length()) {
                    val item = root.optJSONObject(index) ?: continue
                    val id = item.optString("id").trim()
                    val endpoint = item.optString("serverEndpoint").trim()
                    if (id.isEmpty() || endpoint.isEmpty()) {
                        mutated = true
                        continue
                    }
                    if (item.optString("tailnetEnrollmentPayload").trim().isNotEmpty()) {
                        mutated = true
                    }
                    add(
                        BridgeProfile(
                            id = id,
                            bridgeId = item.optString("bridgeId").trim().ifEmpty { null },
                            name = item.optString("name").trim().ifEmpty { endpoint },
                            serverEndpoint = endpoint,
                            authToken = item.optString("authToken").trim().ifEmpty { null },
                            lastUsedAtMillis = item.optLong("lastUsedAtMillis").takeIf { it > 0L },
                            libp2pPeerId = item.optString("libp2pPeerId").trim().ifEmpty { null },
                        ),
                    )
                }
            }.let { profiles ->
                ParsedProfiles(
                    profiles = profiles,
                    mutated = mutated,
                )
            }
        }.getOrDefault(ParsedProfiles(emptyList(), mutated = true))
    }

    private fun persistSanitizedProfilesIfNeeded(parsed: ParsedProfiles): List<BridgeProfile> {
        if (parsed.mutated) {
            val activeProfileId = preferences.getString(KEY_ACTIVE_PROFILE_ID, null)?.trim().orEmpty()
            val nextActiveId = parsed.profiles.firstOrNull { it.id == activeProfileId }?.id ?: parsed.profiles.firstOrNull()?.id
            persistProfiles(parsed.profiles, nextActiveId)
        }
        return parsed.profiles
    }

    private fun persistProfiles(profiles: List<BridgeProfile>, activeProfileId: String?) {
        val serializedProfiles =
            JSONArray().apply {
                profiles.forEach { profile ->
                    put(
                        JSONObject()
                            .put("id", profile.id)
                            .put("bridgeId", profile.bridgeId)
                            .put("name", profile.name)
                            .put("serverEndpoint", profile.serverEndpoint)
                            .put("authToken", profile.authToken)
                            .put("lastUsedAtMillis", profile.lastUsedAtMillis)
                            .put("libp2pPeerId", profile.libp2pPeerId)
                    )
                }
            }
        preferences.edit()
            .putString(KEY_PROFILES_JSON, serializedProfiles.toString())
            .putString(KEY_ACTIVE_PROFILE_ID, activeProfileId)
            .remove(KEY_NAME)
            .remove(KEY_SERVER_ENDPOINT)
            .remove(KEY_AUTH_TOKEN)
            .apply()
    }

    private fun migrateLegacyProfile(): BridgeProfile? {
        val endpoint = preferences.getString(KEY_SERVER_ENDPOINT, null)?.trim().orEmpty()
        if (endpoint.isEmpty()) {
            return null
        }
        val name = preferences.getString(KEY_NAME, null)?.trim().orEmpty().ifEmpty { endpoint }
        val authToken = preferences.getString(KEY_AUTH_TOKEN, null)?.trim()?.ifEmpty { null }
        val profile =
            BridgeProfile(
                id = createProfileId(name, endpoint),
                bridgeId = null,
                name = name,
                serverEndpoint = endpoint,
                authToken = authToken,
                lastUsedAtMillis = System.currentTimeMillis(),
            )
        persistProfiles(listOf(profile), profile.id)
        return profile
    }

    companion object {
        private const val PREFERENCES_NAME = "codex_native_host"
        private const val KEY_NAME = "name"
        private const val KEY_SERVER_ENDPOINT = "server_endpoint"
        private const val KEY_AUTH_TOKEN = "auth_token"
        private const val KEY_PROFILES_JSON = "profiles_json"
        private const val KEY_ACTIVE_PROFILE_ID = "active_profile_id"
    }
}
